git: worktree, add Grep() method for git grep

darkowlzz opened the pull request:
This change implemented grep on worktree with options to **~match ignoring case~, invert match and specify pathspec**. Also, a **commit hash or reference** can be used to specify the worktree to search.

mcuadros commented on options.go:
```diff
@@ -365,3 +365,42 @@ type ListOptions struct {
+// GrepOptions describes how a grep should be performed.
+type GrepOptions struct {
+	// Pattern is the match pattern.
+	Pattern string
```
Itstead of accepting a string, I would prefer a `*regexp.Regexp` with this the people can speed that the process with precompilled expressions.

mcuadros commented on options.go:
```diff
@@ -365,3 +365,42 @@ type ListOptions struct {
+	// Pattern is the match pattern.
+	Pattern string
+	// IgnoreCase enables case insensitive match.
+	IgnoreCase bool
```
If we accept the type instead of the string, this is not need it.

mcuadros commented on worktree.go:
```diff
@@ -711,6 +713,118 @@ func (w *Worktree) Clean(opts *CleanOptions) error {
+	}
+
+	// Iterate through the files and look for any matches.
+	err = fileiter.ForEach(func(file *object.File) error {
```
can we put this code in a private function?

mcuadros commented on worktree_test.go:
```diff
@@ -1317,3 +1317,194 @@ func (s *WorktreeSuite) TestAlternatesRepo(c *C) {
+					TreeName:   "6ecf0ef2c2dffb796033e5a02219af86ec6584e5",
+				},
+			},
+		},
```
I prefer the open and close brace in the same line.

darkowlzz commented:
* Changed `GrepOptions` to accept precompiled regexp objects for `Pattern` and `PathSpec`
* Moved the file iteration and pattern matching code to a separate private function `findMatchInFiles()`
* Changed the table test struct format
* Updated compatibility doc
